onboarding/mlsys/2-autodiff.md

```diff
@@ -0,0 +1,399 @@
+
+# steal a copy of the randomly-initialized weights
+my_l1 = torch.Tensor(l1.weight.detach().clone())
+my_l2 = torch.Tensor(l2.weight.detach().clone())
+my_l1.requires_grad = True
+my_l.requires_grad = True
+
+def my_train_batch(data):
+    pred = my_l2.matmul(my_l1.matmul(data))
+    loss = mse(pred, data)
+    loss.backward()
+
+    #####################
+    # your code goes here
+    #####################
+
+# test your momentum implementation against pytorch's
+for i in range(100):
+    data = torch.rand(5)
+    ptloss = pytorch_train_batch(data)
+    myloss = my_train_batch(data)
+    assert abs(ptloss - myloss) < 0.000001, (
+        f"i = {i}, ptloss = {ptloss}, myloss = {myloss}"
+    )
+```
+
+## TensorFlow
+
+While we do have more PyTorch users than TensorFlow/Keras users, but we develop
+features and field support questions from both frameworks.  We also maintain a
+Keras-based feature in the upstream Horovod project.  So it is important to be
+familiar with both.
+
+### Q7
+
+Train a model using momentum again, this time in TensorFlow.  As before, your
+solution should use low-level Tensor manipulations rather than high-level APIs.
+You might start reading about [tf.Variable](
+https://www.tensorflow.org/guide/variable) and [tf.GradientTape](
+https://www.tensorflow.org/guide/autodiff).
+
+Note that tf.Variable and tf.GradientTape will come up in OSS support, whenever
+users subclass `keras.models.Model` or `keras.metrics.Metric`.
+
+Here's a template to get you started:
+
+```python
+import tensorflow as tf
+from tensorflow.keras import layers, losses, models, optimizers
+
+LR = 0.01
+MOMENTUM = 0.9
+
+# same bogus model as with the pytorch example
+model = models.Sequential()
+model.add(layers.Dense(5, activation=None, use_bias=False, input_shape=(5,)))
+model.add(layers.Dense(5, activation=None, use_bias=False))
+opt = optimizers.SGD(learning_rate=LR, momentum=MOMENTUM)
+model.compile(
+    optimizers.SGD(learning_rate=LR, momentum=MOMENTUM),
+    losses.mean_squared_error,
+)
+
+# Your code here:
+# - copy initial weights
+# - define my_train_batch()
+
+# test your momentum implementation against tensorflow's
+for i in range(100):
+    data = tf.random.uniform((1, 5))
+    tfloss = model.train_on_batch(data, data)
+    myloss = my_train_iter(data)
+    assert abs(tfloss - myloss) < 0.000001, (
+        f"i = {i}, tfloss = {tfloss}, myloss = {myloss}"
+    )
+```
+
+## Optimizers and Large Models
+
+As you have seen, optimizer state is often at least as large as model state,
+even for fairly optimizers strategies like momentum.  This effect is compounded
+with other advanced features like mixed-precision training, where mixed
+precision, where even though model parameters require less memory each, the
+per-parameter optimizer state memory requirement actually grows.
+
+As a result, training large models often means dealing with very large
+optimizer states.  A popular strategy is the [ZeRO Memory Optimization](
```

Comment:
Maybe mention that DeepSpeed is the library which implements the ZeRO strategies.